src/main/java/org/broadinstitute/hellbender/utils/nio/ReadsIterable.java

```diff
@@ -0,0 +1,141 @@
+package org.broadinstitute.hellbender.utils.nio;
+
+import com.google.cloud.storage.contrib.nio.CloudStorageFileSystem;
+import htsjdk.samtools.QueryInterval;
+import htsjdk.samtools.SAMRecord;
+import htsjdk.samtools.SAMRecordIterator;
+import htsjdk.samtools.SamInputResource;
+import htsjdk.samtools.SamReader;
+import htsjdk.samtools.SamReaderFactory;
+import htsjdk.samtools.ValidationStringency;
+import htsjdk.samtools.seekablestream.ByteArraySeekableStream;
+import htsjdk.samtools.seekablestream.SeekableStream;
+import org.broadinstitute.hellbender.utils.gcs.BucketUtils;
+import org.broadinstitute.hellbender.utils.nio.ChannelAsSeekableStream;
+import org.broadinstitute.hellbender.utils.nio.SeekableByteChannelPrefetcher;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+/**
+ * ReadsIterable gives you all the reads for a given genomic interval.
+ *
+ * QueryInterval + header --> iterable SAMRecords
+ */
+public class ReadsIterable implements Iterable<SAMRecord>, Serializable {
+
+    private static final long serialVersionUID = 1L;
+    final String path;
+    final byte[] index;
+    final QueryInterval interval;
+    final boolean removeHeader = true;
+
+    class ReadsIterator implements Iterator<SAMRecord> {
```

Comment:
Are you *really sure* you want to switch to a static inner class? This class is meant to not be static. The conversion would create some redundant, surprising-looking code.

Those issues you saw were perhaps due to all non-static inner class holding a reference to their parent class. This is certainly surprising if you don't know that's going on, and if the parent class isn't designed to serialize can lead to problems.

Here instead, the design calls for the parent class to be serialized, and it's written with that in mind. I would prefer to leave it as is if at all possible.